lib/Dialect/SV/Ops.cpp

```diff
@@ -113,6 +114,54 @@ static void printModportStructs(OpAsmPrinter &p, Operation *,
   p << ')';
 }
 
+//===----------------------------------------------------------------------===//
+// Extract operation.
+//===----------------------------------------------------------------------===//
+
+static ParseResult parseExtractOp(OpAsmParser &p, OperationState &result) {
+  MLIRContext *ctxt = result.getContext();
+  FlatSymbolRefAttr field;
+  OpAsmParser::OperandType extractable;
+  Type inputType;
+  if (p.parseAttribute(field) || p.parseOperand(extractable) || p.parseColon())
+    return failure();
+  result.addAttribute("field", field);
+
+  auto typeLoc = p.getCurrentLocation();
+  if (p.parseType(inputType))
+    return failure();
+
+  if (p.resolveOperand(extractable, inputType, result.operands))
+    return failure();
+
+  // The type specified is the extractable type, so the return type can be
+  // inferred based on that. Said resolution is different depending on the
+  // extractable type.
+  if (auto ifaceTy = inputType.dyn_cast<InterfaceType>()) {
+    SymbolRefAttr ifaceSym = ifaceTy.getInterface();
+    SmallVector<FlatSymbolRefAttr, 4> refs;
+    auto nested = ifaceSym.getNestedReferences();
+    refs.append(nested.begin(), nested.end());
+    refs.push_back(field);
+
+    SymbolRefAttr modportSym =
+        SymbolRefAttr::get(ifaceSym.getRootReference(), refs, ctxt);
+    auto resultPort = InterfaceModportType::get(ctxt, modportSym);
+    result.addTypes({resultPort});
+    return success();
+  } else {
+    p.emitError(typeLoc, llvm::formatv("Expected extractable type, got '{0}'",
+                                       inputType));
+    return failure();
+  }
+}
+
+static void printExtractOp(OpAsmPrinter &p, ExtractOp &op) {
+  p << "sv.extract " << op.fieldAttr() << " " << op.extractable() << " ";
+  p.printOptionalAttrDict(op.getAttrs(), /* elidedAttrs */ {"field"});
+  p << ": " << op.extractable().getType();
```

Comment:
I was expecting the result type to be printed here when I read the test case. Isn't it kind of redundant to print the input type as opposed to the result type? When the extracted value gets used, for example in `rtl.instance`, it's the result type that is printed by the using op.

Reply:
I went with the input type because the output type can be inferred from the input type, but not the other way around so you'd end up having to specify both the input and output type, which is redundant. I prefer the syntax to be as non-redundant as possible.

Reply:
I'm still not quite following why the input type can't be inferred from the output, or really where we would need to infer the input type at all. Is it not explicitly `op.extractable().getType()`? 

I agree it shouldn't be redundant, but I guess I'm still not getting why the assembly format can't just put the result type here. Looking also at this [comment](https://github.com/llvm/circt/pull/294#discussion_r535527190), what prevents the declarative syntax from just printing the result type. Does it not let you just put `type($result)`?

Reply:
> I'm still not quite following why the input type can't be inferred from the output, or really where we would need to infer the input type at all. Is it not explicitly op.extractable().getType()?

Say you have: `struct s { i32 i, u32 j }`. When you `%i = sv.extract @i %sinst : i32` you can't infer (without context way) that the `i32` comes from the `struct s`, which is the input. Same for interface types.

>  Does it not let you just put type($result)?

It also requires `type($extractable)` since the logic to determine the type(s) has to be in the parser code. Though come to think of it, I could probably use a custom directive to add that logic.

Reply:
I see now. I didn't look into using a custom, directive, but that could be a nice middle-ground. IMHO, something like this assembly format isn't too bad:

```
let assemblyFormat = [{
    $extractable $field attr-dict `:` functional-type($extractable, $result)
}];
```

Reply:
Turns out the custom directive doesn't supply the information necessary to pull this off. Bummer man.